### PR TITLE
Feature/BulletReplacement

### PR DESCRIPTION
**Pull request by toryjleo**
https://trello.com/c/k9eESJhW/79-gun-system-bullets
- Bullet refactor
- Add HitScan class
- Create GunTester scene & script

### CODE REVIEW

File: Assets/Code/Editor/CustomInspector/Scripts/GunStats.cs [hunk 1]
```diff
@@ -176,7 +178,9 @@ private void GeneratedStats(EditorObject.GunStats gunStats)
             {
                 EditorGUILayout.LabelField("Time To Cool After Overheat: " + timeToCoolAfterOverheat + " seconds");
             }
+
+            // TODO: Make a visual of object so you can see scale applied in realtime
```
Comment (toryjleo) [inline Assets/Code/Editor/CustomInspector/Scripts/GunStats.cs:182]: Look into this

File: Assets/Code/Scripts/Bullets/HitScan.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        int n = Mathf.Min(gunStats.BulletPenetration + 1, hits.Length);
+
+        for (int i = 0; i < n; i++)
+        {
+            RaycastHit hit = hits[i];
+            Debug.Log(hit.transform.name);
+
+            if ((hit.transform.tag == "Enemy" && gunStats.IsPlayerGun) ||
+                (hit.transform.tag == "Player" && !gunStats.IsPlayerGun))
```
Comment (LiamMillerSE) [inline Assets/Code/Scripts/Bullets/HitScan.cs:44]: Do we want a player "hit" to reduce the pen number? I think that if the bullet hits the player it's probably a clipping issue

File: Assets/Code/Scripts/Bullets/HitScan.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        Health otherHealth = other.GetComponentInChildren<Health>();
+        if (otherHealth == null)
+        {
+            Debug.LogError("Object does not have Health component: " + other.name);
+        }
+        else
+        {
+            otherHealth.TakeDamage(gunStats.DamageDealt);
+
+            // TODO: instantiate and play impact effect particlesystem at hit.point and rotate to normal
```
Comment (LiamMillerSE) [inline Assets/Code/Scripts/Bullets/HitScan.cs:73]: Is this in the backlog?

**Review by LiamMillerSE — APPROVED** [review]
Consider comments, but approved :)

File: Assets/Code/Scripts/Bullets/HitScan.cs (hunk 2 above)
Comment (toryjleo) [reply Assets/Code/Scripts/Bullets/HitScan.cs:44]: What if an enemy hits a player or the player shoots a shot that then hits the player?

**Pull request merged**